pkg/build/controller/buildpod/controller.go

```diff
@@ -0,0 +1,448 @@
+	}
+
+	return obj.(*kapi.Pod), nil
+}
+
+func (bc *BuildPodController) updatePod(old, cur interface{}) {
+	// A periodic relist will send update events for all known pods.
+	curPod := cur.(*kapi.Pod)
+	oldPod := old.(*kapi.Pod)
+	if curPod.ResourceVersion == oldPod.ResourceVersion {
+		return
+	}
+	if isBuildPod(curPod) {
+		bc.enqueuePod(curPod)
+	}
+}
+
+func (bc *BuildPodController) deletePod(obj interface{}) {
+	pod, ok := obj.(*kapi.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone: %+v", obj))
+			return
+		}
+		pod, ok = tombstone.Obj.(*kapi.Pod)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a pod: %+v", obj))
+			return
+		}
+	}
+	if isBuildPod(pod) {
+		err := bc.HandleBuildPodDeletion(pod)
+		utilruntime.HandleError(err)
+	}
+}
+
+func (bc *BuildPodController) addBuild(obj interface{}) {
+	build := obj.(*buildapi.Build)
+	bc.checkBuildPodDeletion(build)
+}
+
+func (bc *BuildPodController) updateBuild(old, cur interface{}) {
+	build := cur.(*buildapi.Build)
+	bc.checkBuildPodDeletion(build)
+}
+
+func (bc *BuildPodController) checkBuildPodDeletion(build *buildapi.Build) {
+	switch {
+	case buildutil.IsBuildComplete(build):
+		glog.V(5).Infof("checkBuildPodDeletion: ignoring build %s/%s because it is complete", build.Namespace, build.Name)
+		return
+	case build.Status.Phase == buildapi.BuildPhaseNew:
+		glog.V(5).Infof("checkBuildPodDeletion: ignoring build %s/%s because it is new", build.Namespace, build.Name)
+		return
+	case build.Spec.Strategy.JenkinsPipelineStrategy != nil:
+		glog.V(5).Infof("checkBuildPodDeletion: ignoring build %s/%s because it is a pipeline build", build.Namespace, build.Name)
+		return
+	}
+	_, err := bc.podStore.Pods(build.Namespace).Get(buildapi.GetBuildPodName(build))
+
+	// The only error that can currently be returned is a NotFound error, but adding a check
+	// here just in case that changes in the future
+	if err != nil && errors.IsNotFound(err) {
+		bc.enqueuePodNotFound(build.Namespace, buildapi.GetBuildPodName(build), build.Name)
+	}
+	if err != nil {
+		utilruntime.HandleError(err)
+	}
+}
+
+func (c *BuildPodController) enqueuePod(pod *kapi.Pod) {
+	key, err := kcontroller.KeyFunc(pod)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", pod, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *BuildPodController) enqueuePodNotFound(namespace, name, buildName string) {
+	key := podNotFoundKey{
+		Namespace: namespace,
+		PodName:   name,
+		BuildName: buildName,
+	}
+	c.queue.Add(key)
+}
+
+func (bc *BuildPodController) handleErr(err error, key interface{}, pod *kapi.Pod) {
+	if err == nil {
+		bc.queue.Forget(key)
+		return
+	}
+
+	if strategy.IsFatal(err) {
+		glog.V(2).Infof("Will not retry fatal error for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		bc.queue.Forget(key)
+		return
+	}
```

Comment:
is there any way to throw a fatal error in this code?  I'm not seeing one.

what conditions would we want to consider as fatal for handling a pod event?